### ISSUE TITLE
[feature] Add option "--ignore-binaries" to conan install or --build=notincache

### ISSUE THREAD

**Issue by fukanchik** [conan-io/conan#9994]
For a subset of packages from https://center.conan.io we get 
```
undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
```
on our build machines. 

To continue the build we use option `--build=<problematic-package>` however this is very annoying for bigger packages which are already rebuilt and available from the local cache.

We understand that glibc version won't be added as a first-class setting into conan. Therefore we request feature to only rebuild specific dependency only when it is not available in the local cache.

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @fukanchik 

If you need to re-build your own binaries for an older glibc version, I would recommend another approach:

- Add a new subsetting to your ``Linux`` ``os`` in ``settings.yml``, called something like ``distro = ["mydistro"]``
- Define ``os.distro = mydistro`` in your profiles
- This will compute a new package_id, different to the ConanCenter ones
- This work consistently in the cache, and also if you upload your own binaries to your own private Artifactory (recommended, to avoid rebuilding from source, specially if they are big as you say, recall that ArtifactoryCE is completely free). 

This solution is more generic and powerful, because it doesn't limit to builds in the cache, and can leverage the whole Conan power reusing binaries.

**Comment by fukanchik** [comment]
Thanks, no, we do not need to rebuild my own binaries as we only ship our conan package as a source. Ticket is about the binaries which are downloaded as a dependency from center.conan.io and therefore are not under our control.

Examples which have glibc problems for our users are protobuf and grpc.

**Comment by memsharded** [comment]
> Thanks, no, we do not need to rebuild my own binaries as we only ship our conan package as a source.

But you are building your own binaries, in the local cache, and not using the binaries from ConanCenter, right? So you definitely need to build your own binaries.

You don't need to ship your binaries, you can totally ship your package as source. Still, explicitly modeling that the binary is different from the ConanCenter one is a better way to do it. Lets say that you ship your Conan package as source, but the consumers of that package want to build and store their binaries in their own server to avoid building from source. They will be able to do it.

**Comment by fukanchik** [comment]
In our own institution it can probably be solved by banning center.conan.io and using internal artifactory. However users build our software in all sorts of local setups and cloud environments. In many cases users are not very experienced and we would like to not ask them to install local artifactory just to build our package.

settings.xml -- if I understand it correctly adding `distro` into our settings.xml would not affect grpc and protobuf downloaded in binary form from conancenter and distro would be None for these?

**Comment by memsharded** [comment]
> In our own institution it can probably be solved by banning center.conan.io and using internal artifactory. However users build our software in all sorts of local setups and cloud environments. In many cases users are not very experienced and we would like to not ask them to install local artifactory just to build our package.

Yes, that is not an issue, they doesn't need to install Artifactory to build your package. Just in case some of they want to (and this is expected), they will be able to do it and be able to succesfully reuse binaries. The proposed approach doesn't have Artifactory as a requirement, but it is enabling both cases, both using source-only and building in the cache, but also to be able to use a server to store the binaries if they want to.

> settings.xml -- if I understand it correctly adding distro into our settings.xml would not affect grpc and protobuf downloaded in binary form from conancenter and distro would be None for these?

Yes, you can for example add ``distro = [None, "mycustom"]``. ``None`` will be the default one, and will match package_ids from ConanCenter. For those packages that you want to explicitly build and manage a different binary you select the custom one. Your own profile or command line would be relatively simple, recall that you can also set ``settings`` per package:

```
[settings]
grpc:os.distro=mycustom
protobuf:os.distry=mycustom
```

Also recall that you can put profiles and settings.yml in a git repo, and let users do a ``conan config install`` of it, so this enables also a way to communicate the intention and the different binaries existence.

**Comment by vakatov** [comment]
> put profiles and settings.yml in a git repo, and let users do a `conan config install`

We'd very much like to avoid additional ad hoc tuneups on the user side. Rather, things should "just work".

We'd rather do all required adjustments on our side, invisibly for the users. E.g. in this case we _know_ that `PROTOBUF/GRPC` default (i.e. from https://center.conan.io) pre-built binaries require `GLIBC_2.2.5` -- and therefore  we can adjust our code to only use binaries when the latter is available on the end-user host.

Naturally, we can just as well call `conan config install` from our build process to use the suggested `mycustom` solution... but that is not nice -- because that'll affect user's _global_ `conan` settings. Another problem with the `mycustom` solution is that we don't really want to maintain alternative binaries; instead we'd prefer to just have such binaries built locally (on the end-user host).

One other issue is that with `mycustom` solution it may become difficult to easily and robustly manage the build/nobuild modes for different packages and different build environments.

This said, I see how our situation can be thought of as an esoteric corner case -- with outdated `LIBC `on our side. OTOH, unless corner cases are taken care of the whole framework can become less usable. Ignoring too many corner cases can lead to a situation where these cases -- while each of these is relatively rare -- can all together break things way too often with all these hundreds of interdependent packages in the ecosystem.

PS: Not having `LIBC` (or, more generally, something that reflects critical system runtime environment) as a standard setting obviously creates fertile grounds for unnecessary problems and confusion (for the laggards like ourselves but possibly for, well, other laggards in other big institutions).

**Comment by memsharded** [comment]
Conan 2 added the ``--recipes-only`` to ``remote`` definition, that achieves the requested feature.
Closing as solved in Conan 2, thanks for the feedback.

**Comment by vakatov** [comment]
Thanks!
